### PR TITLE
Add Attribute to Group Mapper for SAML IDP

### LINKED ISSUE #12950: Implement "advanced claim to group" mapping for SAML
### Description

Currently, Keycloak supports mappers for Identity Providers.  Recently the ability to map an "advanced claim to [a] group" [was added](https://github.com/keycloak/keycloak/pull/8467) as a new Mapper Type, but only implemented for "OpenID Connect v1.0" Identity Providers.  Adding the same for SAML Identity Providers would help to keep feature parity between the 2 common standards.

### Discussion

_No response_

### Motivation

Many identity providers still utilize SAML and would like to map in groups.  It would therefore be helpful to implement the same mapper for SAML that was added for OpenID Connect.

### Details

_No response_

### PR DESCRIPTION
**Pull request by bernard-denis**
Add feature to map attribute to a group with saml IDP Provider 

Tested with Azure AD Saml IDP locally 

Closes #12950

### CODE REVIEW

File: services/src/main/java/org/keycloak/broker/saml/mappers/AbstractAttributeToGroupMapper.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.IdentityProviderMapperModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.utils.KeycloakModelUtils;
+
+/**
+ * Abstract class that handles the logic for importing and updating brokered users for all mappers that map a SAML
+ * attribute into a {@code Keycloak} role.
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/saml/mappers/AbstractAttributeToGroupMapper.java:32]: The comment is still the same from `AbstractAttributeToRoleMapper`.

File: services/src/main/java/org/keycloak/broker/saml/mappers/AbstractAttributeToRoleMapper.java [hunk 2]
```diff
@@ -87,8 +87,7 @@ public void updateBrokeredUser(KeycloakSession session, RealmModel realm, UserMo
      *
      * @param realm a reference to the realm.
      * @param mapperModel a reference to the {@link IdentityProviderMapperModel} containing the configured role.
-     * @return the {@link RoleModel} that corresponds to the mapper model role or {@code null}, if the role could not be
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/saml/mappers/AbstractAttributeToRoleMapper.java:90]: Could you please remove this change?

**Review by pedroigor — CHANGES_REQUESTED** [review]
Thanks, @bernard-denis.

Do you mind looking at this [guide](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md) and following the first steps there to contribute to the project?

Basically, you should create an issue and link it to this PR as per the guide above.

The mapper you are proposing should enable and make it easier to assign federated users to groups dynamically. It should be a common use case.

We also need an integration test for it. It can be done similarly to `https://github.com/keycloak/keycloak/blob/f80a8fbed0940f28e8a23ebcdc5469c7299c0457/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToRoleMapperTest.java#L21`. The same mapper that grants roles based on attributes is being tested by that test class.

File: services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToGroupMapper.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.broker.saml.mappers;
+
+/**
+ * @author <a href="mailto:denis.bernard@avanade.com">Denis Bernard</a>
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToGroupMapper.java:21]: Not here?

**Review by pedroigor — CHANGES_REQUESTED** [review]

**Comment by bernard-denis** [comment]
Thank @pedroigor for your review. 

I followed the contributing guide as well I understand it. 
I tagged my branch the issue number like explain,
I tagged the commit with the issue already created by community 
I didn't see how to link more my commit to the issue #12950 that I did. 

Sorry for the mistake test. 

As you can see, I build this function based on the current SAML Role mapper and the OIDC Group Mapper. 
and I didn't see this kind of test in the OIDC pull request (maybe I have miss read it). My bad to forget it
I'm not a developer, but I need this feature for my current customer, so I did my best for the contribution :) 

Thank for checking my work. I'll make the correction early.

File: services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToGroupMapper.java (hunk 3 above)
Comment (pedroigor) [reply services/src/main/java/org/keycloak/broker/saml/mappers/AdvancedAttributeToGroupMapper.java:21]: Sorry, just like other classes you are adding, the `@author` should be right above the class declaration.

**Comment by pedroigor** [comment]
@bernard-denis No worries, we appreciate your time on this. It is a useful mapper.

If you need help with the test, please let me know. That should be the last missing bit.

**Comment by bernard-denis** [comment]
@pedroigor It seems good now. 
I build the test based on the SAML and OIDC existing test and made the correction asked
the CI task ran successfully (on my self hosted runner) the build and test (some part failed for missing dependency in my runner :) ).

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToGroupMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                                     boolean areClaimsOrAttributeValuesRegexes, IdentityProviderMapperSyncMode syncMode, String groupValue) {
+        IdentityProviderMapperRepresentation advancedAttributeToGroupMapper = new IdentityProviderMapperRepresentation();
+        advancedAttributeToGroupMapper.setName("advanced-attribute-to-group-mapper");
+        advancedAttributeToGroupMapper.setIdentityProviderMapper(AdvancedAttributeToGroupMapper.PROVIDER_ID);
+        advancedAttributeToGroupMapper.setConfig(ImmutableMap.<String, String> builder()
+                .put(IdentityProviderMapperModel.SYNC_MODE, syncMode.toString())
+                .put(AdvancedAttributeToGroupMapper.ATTRIBUTE_PROPERTY_NAME, claimsOrAttributeRepresentation)
+                .put(AdvancedAttributeToGroupMapper.ARE_ATTRIBUTE_VALUES_REGEX_PROPERTY_NAME,
+                        Boolean.valueOf(areClaimsOrAttributeValuesRegexes).toString())
+                .put(ConfigConstants.GROUP, groupValue)
```
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToGroupMapperTest.java:55]:
```suggestion
                .put(ConfigConstants.GROUP, MAPPER_TEST_GROUP_PATH)
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToGroupMapperTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        createAdvancedGroupMapper(ATTRIBUTES, false,KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2);
+        createUserInProviderRealm(ImmutableMap.<String, List<String>> builder()
+                .put(ATTRIBUTE_TO_MAP_FRIENDLY_NAME, ImmutableList.<String> builder().add("value 1").build())
+                .put(KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2,
+                        ImmutableList.<String> builder().add("value 2").build())
+                .build());
+
+        logInAsUserInIDPForFirstTime();
+        UserRepresentation user = findUser(bc.consumerRealmName(), bc.getUserLogin(), bc.getUserEmail());
+        assertThatUserHasBeenAssignedToGroup(user, KcOidcBrokerConfiguration.ATTRIBUTE_TO_MAP_NAME_2);
```
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlAdvancedAttributeToGroupMapperTest.java:75]:
```suggestion
        assertThatUserHasBeenAssignedToGroup(user, MAPPER_TEST_GROUP_PATH);
```

**Review by pedroigor — COMMENTED** [review]
@bernard-denis Did some suggestions in order to make your test method to pass.

I think it will fail because you are not configuring the mapper to point to a valid group (`MAPPER_TEST_GROUP_PATH`)

To run the test, run the following command:

```
 mvn -f testsuite/integration-arquillian/pom.xml -Pauth-server-quarkus clean install -Dtest=KcCustomOidcBrokerTest
```

After that, you can run from your IDE too by enabling the `auth-server-quarkus-embedded` maven profile and then running the test class.

**Comment by pedroigor** [comment]
@bernard-denis Here is the changes I did https://github.com/pedroigor/keycloak/tree/tmp-broker.

We also need to remove the mapper after running the test otherwise it will mess up other test methods.

Not so good, but this area is a bit odd and we need to refactor some brokering tests ...

**Comment by pedroigor** [comment]
@bernard-denis In order to be able to implement the test, I had to make some changes completely out of the scope. To avoid spending more of your time, I've sent https://github.com/keycloak/keycloak/pull/16840 with your changes plus those to the test.

I'm closing this one in favor of https://github.com/keycloak/keycloak/pull/16840.

**Comment by obscuredByCloud** [comment]
Hello, I am trying to map group idp attribuite to a group, this seems to work but it doesn't join the group.  
keycloak is managing the group and the saml IDP authenicates and sends back the name and group and I see the user attribte group is mapped, but when I look at the user, it only gives option to click on join, it doesn't auto-join the group. I'm probably missing something, or do I have to add code to do this?

**Pull request closed without merging**